### PR TITLE
Add serialization (nostd)

### PR DESCRIPTION
**Pull request by kampersanda**
This PR added serialization/deserialization to/from byte sequence, while keeping no_std.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -52,6 +57,49 @@ pub trait Statistics {
     }
 }
 
+/// Trait to serialize/deserialize the data structure.
+pub trait Serializer {
```
[the remark this answers is not in the document]
Comment (kampersanda) [reply src/lib.rs]: I removed the trait of `Serializer`. https://github.com/daac-tools/crawdad/pull/14/commits/d8ea62278084a430e3b1053400ed5105efbcaa3e

`Statistics` will be removed in another PR.

File: src/mptrie.rs [hunk 2]
```diff
@@ -97,6 +99,89 @@ impl MpTrie {
...
+    /// ```
+    pub fn deserialize_from_slice(source: &[u8]) -> (Self, &[u8]) {
+        let (mapper, mut source) = CodeMapper::deserialize_from_slice(source);
+        let nodes = {
+            let len = u32::from_le_bytes(source[..4].try_into().unwrap()) as usize;
+            source = &source[4..];
+            let mut nodes = Vec::with_capacity(len);
+            for _ in 0..len {
+                nodes.push(Node::deserialize(
+                    source[..size_of::<Node>()].try_into().unwrap(),
```
Comment (vbkaisetsu) [inline src/mptrie.rs:158]: There is no guarantee that the size of the serialized data is the same as the size of the structure.
```suggestion
                    source[..8].try_into().unwrap(),
```
or
```suggestion
                    source[..Node::io_bytes()].try_into().unwrap(),
```

File: src/mptrie.rs [hunk 3]
```diff
@@ -97,6 +99,89 @@ impl MpTrie {
...
+                ));
+                source = &source[size_of::<Node>()..];
```
Comment (vbkaisetsu) [inline src/mptrie.rs:160]:
```suggestion
                source = &source[8..];
```

File: src/mptrie.rs [hunk 4]
```diff
@@ -299,6 +384,15 @@ impl Statistics for MpTrie {
             + self.tails.len() * core::mem::size_of::<u8>()
     }
 
+    fn io_bytes(&self) -> usize {
+        self.mapper.io_bytes()
+            + self.nodes.len() * size_of::<Node>()
```
Comment (vbkaisetsu) [inline src/mptrie.rs:389]: ditto

File: src/mptrie.rs [hunk 5]
```diff
@@ -299,6 +384,15 @@ impl Statistics for MpTrie {
...
+            + size_of::<u32>()
+            + self.tails.len() * core::mem::size_of::<u8>()
```
Comment (vbkaisetsu) [inline src/mptrie.rs:391]:
```suggestion
            + self.tails.len() * size_of::<u8>()
```

File: src/trie.rs [hunk 6]
```diff
@@ -88,6 +90,67 @@ impl Trie {
...
+    /// ```
+    pub fn deserialize_from_slice(source: &[u8]) -> (Self, &[u8]) {
+        let (mapper, mut source) = CodeMapper::deserialize_from_slice(source);
+        let nodes = {
+            let len = u32::from_le_bytes(source[..4].try_into().unwrap()) as usize;
+            source = &source[4..];
+            let mut nodes = Vec::with_capacity(len);
+            for _ in 0..len {
+                nodes.push(Node::deserialize(
+                    source[..size_of::<Node>()].try_into().unwrap(),
```
Comment (vbkaisetsu) [inline src/trie.rs:145]: ditto

File: src/trie.rs [hunk 7]
```diff
@@ -88,6 +90,67 @@ impl Trie {
...
+                ));
+                source = &source[size_of::<Node>()..];
```
Comment (vbkaisetsu) [inline src/trie.rs:147]: ditto

File: src/trie.rs [hunk 8]
```diff
@@ -247,7 +310,11 @@ impl Trie {
 
 impl Statistics for Trie {
     fn heap_bytes(&self) -> usize {
-        self.mapper.heap_bytes() + self.nodes.len() * core::mem::size_of::<Node>()
+        self.mapper.heap_bytes() + self.nodes.len() * size_of::<Node>()
```
Comment (vbkaisetsu) [inline src/trie.rs:313]: ditto

File: src/trie.rs [hunk 9]
```diff
@@ -247,7 +310,11 @@ impl Trie {
...
+    }
+
+    fn io_bytes(&self) -> usize {
+        self.mapper.io_bytes() + self.nodes.len() * size_of::<Node>() + size_of::<u32>()
```
Comment (vbkaisetsu) [inline src/trie.rs:317]: ditto

File: src/mapper.rs [hunk 10]
```diff
@@ -46,4 +46,39 @@ impl CodeMapper {
     pub fn heap_bytes(&self) -> usize {
         self.table.len() * core::mem::size_of::<u32>()
     }
+
+    #[inline]
+    pub fn io_bytes(&self) -> usize {
+        self.heap_bytes() + core::mem::size_of::<u32>() * 2
```
Comment (vbkaisetsu) [inline src/mapper.rs:52]: `io_bytes()` should not be calculated using `heap_bytes()`.
Could you calculate this from scratch?

**Comment by kampersanda** [comment]
@vbkaisetsu 

Thank you for your suggestions!
I fixed all. https://github.com/daac-tools/crawdad/pull/14/commits/6aad4ab8808b973bec7c980754e0c4188e586dfd

File: src/trie.rs (hunk 8 above)
Comment (vbkaisetsu) [reply src/trie.rs:313]: @kampersanda reminder

**Review by vbkaisetsu — COMMENTED** [review]
.

File: src/trie.rs (hunk 8 above)
Comment (kampersanda) [reply src/trie.rs:313]: @vbkaisetsu 
I've already fixed it, but Files Changed is not updated...
See https://github.com/daac-tools/crawdad/blob/add-serialize/src/trie.rs#L313

Comment (vbkaisetsu) [reply src/trie.rs:313]: Ah... I commented to heap_bytes(). It's my mistake.

**Review by vbkaisetsu — APPROVED** [review]

**Pull request merged**